### DISCUSSION TITLE
WSL Bridge - Packet Diagram

### DISCUSSION THREAD

**Discussion by penguinpages** [microsoft/WSL#8224] [General]
### Version

Version 10.0.22000.556 (Windows 11)

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.10.102.1

### Distro Version

Ubuntu 20.04

### Repro Steps

Related thread: https://docs.microsoft.com/en-us/answers/questions/788964/wsl-with-bridging-windows-11.html

Question: 

WSL. 
1) Does it use bridge to physical NIC and so Windows 11 "Bridge" is means to link from WSL virtual NIC to the physical device
2) Or.. Does it link the WSL virtual nic to HyperV switch.

TOday I can get it to work with HyperV switch "WSL" binding to Gb NIC (with mgmt so base OS works), then enable bridging WSL virtual adapter to Gb NIC.. Total hack job but.. WSL then routes and so does OS.  But.. this  is  1) Brittle and has to be recomfigured each reboot  2) Is more of poking in the dark then tactically setting up packet paths   3) Does not afford configuration profiles where how I connect (Gb, vs Wifi, vs VPN Virtual adapter) based on where I have to work with laptop.

If someone can explain the data path and connection logic,  as well as to "WSL is hidden HyperV guest" and if so..  WHy are we hiding this?  What can we do at a command prompt to effect the environment?

![Windows11_WSL_HV_Bridging_Wifi](https://user-images.githubusercontent.com/17153876/160697827-b56b9077-a5d2-440f-92a9-a63a4251b4cf.gif)

![Windows11_WSL_HV_Bridging_Wifi](https://user-images.githubusercontent.com/17153876/160697846-71b5e2bb-39a0-4625-8a88-0dcfc5bfa791.jpg)

### Expected Behavior

I link WSL to a logical interface:

"Bridge" adapter -> to do L2 bridging to a given NIC, or,  Bridge to hyperV switch, which then offloads communication under known HyperV connection setup.... such as link to interface, or internal etc..

I want to run a command, and get back "how is WSL connected" .. the vNIC is listed, and return back is to what device / NIC that the Windows OS controls.

### Actual Behavior

It is hidden.
No commands exist to list what the virtual WSL host is connecting to
No means to track down links and so configuration
No means to change profile of bindings as host connections change: Ex:  three profiles based on which "NIC" I have to link the WSL intance to.

### Diagnostic Logs

Attached are images of trying to poke in the dark... to get things to work, as well as control with best practices.

**Comment by penguinpages** [comment]
It was suggested to view WSL switch configuruation and mapping to use commands with mixed result and maybe others can post means to map.  I expected to see mapping of MAC of vNIC of Ubuntu Instance to map to MAC of some defined bridge interface.. but did not see that.  

Windows Admin Powershell
`PS C:\WINDOWS\system32> ovn-nbctl
ovn-nbctl : The term 'ovn-nbctl' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ ovn-nbctl
+ ~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (ovn-nbctl:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

PS C:\WINDOWS\system32> get-hnsnetwork


ActivityId             : B3FFFB67-59A6-4954-9CBD-7048C9C33DBD
AdditionalParams       :
CurrentEndpointCount   : 0
Extensions             : {@{Id=E7C3B2F0-F3C5-48DF-AF2B-10FED6D72E7A; IsEnabled=False; Name=Microsoft Windows Filtering
                         Platform}, @{Id=E9B59CFA-2BE1-4B21-828F-B6FBDBDDC017; IsEnabled=False; Name=Microsoft Azure
                         VFP Switch Extension}, @{Id=430BDADD-BAB0-41AB-A369-94B67FA5BE0A; IsEnabled=True;
                         Name=Microsoft NDIS Capture}}
Flags                  : 11
GatewayMac             : 00-15-5D-64-8A-00
Health                 : @{AddressNotificationMissedCount=0; AddressNotificationSequenceNumber=1;
                         DHCPNotificationMissedCount=0; DNSCacheNotificationMissedCount=0;
                         DNSCacheNotificationSequenceNumber=0; DNSNotificationMissedCount=0;
                         DNSNotificationSequenceNumber=0; InterfaceNotificationMissedCount=0;
                         InterfaceNotificationSequenceNumber=0; LastErrorCode=0; LastUpdateTime=132926056728747878;
                         MacAddressNotificationMissedCount=0; MacAddressNotificationSequenceNumber=0;
                         NeighborNotificationMissedCount=0; NeighborNotificationSequenceNumber=0;
                         RouteNotificationMissedCount=0; RouteNotificationSequenceNumber=0;
                         XlatNotificationMissedCount=0; XlatNotificationSequenceNumber=0}
ID                     : C08CB7B8-9B3C-408E-8E30-5E16A3AEB444
IPv6                   : False
LayeredOn              : D7F15566-9910-4732-A7DB-E7359B7E6B77
MacPools               : {@{EndMacAddress=00-15-5D-6A-BF-FF; StartMacAddress=00-15-5D-6A-B0-00}}
MaxConcurrentEndpoints : 0
Name                   : Default Switch
NatName                : ICSC5448807-C36F-41D7-A303-8871B2861C75
Policies               : {}
State                  : 1
Subnets                : {@{AdditionalParams=; AddressPrefix=172.20.96.0/20; Flags=0; GatewayAddress=172.20.96.1;
                         Health=; ID=B3B802D5-C4FE-4F76-8EA1-4261ED4316FA; IpSubnets=System.Object[]; ObjectType=5;
                         Policies=System.Object[]; State=0}}
SwitchGuid             : C08CB7B8-9B3C-408E-8E30-5E16A3AEB444
SwitchName             : Default Switch
TotalEndpoints         : 0
Type                   : ICS
Version                : 60129542144
Resources              : @{AdditionalParams=; AllocationOrder=2; Allocators=System.Object[];
                         CompartmentOperationTime=0; Flags=0; Health=; ID=B3FFFB67-59A6-4954-9CBD-7048C9C33DBD;
                         PortOperationTime=0; State=1; SwitchOperationTime=0; VfpOperationTime=0;
                         parentId=F0E875DB-992B-48BD-8F3B-239AE5D652F3}

ActivityId             : 2A1EFB9A-258A-40EE-8201-044D4ABB5E40
AdditionalParams       :
CurrentEndpointCount   : 0
DNSServerList          : 172.23.160.1
Extensions             : {@{Id=E7C3B2F0-F3C5-48DF-AF2B-10FED6D72E7A; IsEnabled=False; Name=Microsoft Windows Filtering
                         Platform}, @{Id=E9B59CFA-2BE1-4B21-828F-B6FBDBDDC017; IsEnabled=False; Name=Microsoft Azure
                         VFP Switch Extension}, @{Id=430BDADD-BAB0-41AB-A369-94B67FA5BE0A; IsEnabled=True;
                         Name=Microsoft NDIS Capture}}
Flags                  : 9
GatewayMac             : 00-15-5D-03-7D-B7
Health                 : @{AddressNotificationMissedCount=0; AddressNotificationSequenceNumber=0;
                         DHCPNotificationMissedCount=0; DNSCacheNotificationMissedCount=0;
                         DNSCacheNotificationSequenceNumber=0; DNSNotificationMissedCount=0;
                         DNSNotificationSequenceNumber=0; InterfaceNotificationMissedCount=0;
                         InterfaceNotificationSequenceNumber=0; LastErrorCode=0; LastUpdateTime=132926062586346043;
                         MacAddressNotificationMissedCount=0; MacAddressNotificationSequenceNumber=0;
                         NeighborNotificationMissedCount=0; NeighborNotificationSequenceNumber=0;
                         RouteNotificationMissedCount=0; RouteNotificationSequenceNumber=0;
                         XlatNotificationMissedCount=0; XlatNotificationSequenceNumber=0}
ID                     : B95D0C5E-57D4-412B-B571-18A81A16E005
IPv6                   : False
IsolateSwitch          : True
LayeredOn              : 66A34A8A-694B-4C49-A07A-1396CBA640E2
MacPools               : {@{EndMacAddress=00-15-5D-03-7F-FF; StartMacAddress=00-15-5D-03-70-00}}
MaxConcurrentEndpoints : 1
Name                   : WSL
NatName                : ICSEF5CE00F-95A8-4370-A1D8-DA4B257783F5
Policies               : {}
State                  : 1
Subnets                : {@{AdditionalParams=; AddressPrefix=172.23.160.0/20; Flags=0; GatewayAddress=172.23.160.1;
                         Health=; ID=C81DF97C-004B-46CA-B75A-4A81746D13EA; IpSubnets=System.Object[]; ObjectType=5;
                         Policies=System.Object[]; State=0}}
TotalEndpoints         : 10
Type                   : ICS
Version                : 60129542144
Resources              : @{AdditionalParams=; AllocationOrder=2; Allocators=System.Object[];
                         CompartmentOperationTime=0; Flags=0; Health=; ID=2A1EFB9A-258A-40EE-8201-044D4ABB5E40;
                         PortOperationTime=0; State=1; SwitchOperationTime=0; VfpOperationTime=0;
                         parentId=BC684798-A620-4D50-A213-1023FEE347EE}



PS C:\WINDOWS\system32>`


WSL Instance side
`PS C:\GitHub\cluster_devops\cluster_devops> wsl
[sudo] password for ansible: 
ansible@laptop:/mnt/c/GitHub/blah/cluster_devops$ ip ad
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: bond0: <BROADCAST,MULTICAST,MASTER> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 06:06:dc:99:ab:84 brd ff:ff:ff:ff:ff:ff
3: dummy0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 26:b1:da:1e:91:e9 brd ff:ff:ff:ff:ff:ff
4: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
5: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/sit 0.0.0.0 brd 0.0.0.0
6: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:15:5d:03:75:5f brd ff:ff:ff:ff:ff:ff
    inet 172.16.100.133/24 scope global eth0
       valid_lft forever preferred_lft forever
ansible@laptop:/mnt/c/GitHub/blah/cluster_devops$`


List of NICs from OS perspective
`PS C:\WINDOWS\system32> ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : laptop
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : local

Ethernet adapter Ethernet:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . : myhome.local
   Description . . . . . . . . . . . : Intel(R) Ethernet Connection (6) I219-LM
   Physical Address. . . . . . . . . : 38-22-E2-4B-03-F7
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Unknown adapter OpenVPN Wintun:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Wintun Userspace Tunnel
   Physical Address. . . . . . . . . :
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

Ethernet adapter vEthernet (Default Switch):

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter
   Physical Address. . . . . . . . . : 00-15-5D-51-74-30
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::7f:7942:138b:3052%71(Preferred)
   IPv4 Address. . . . . . . . . . . : 172.20.96.1(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.240.0
   Default Gateway . . . . . . . . . :
   DHCPv6 IAID . . . . . . . . . . . : 1191187805
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-27-33-B4-AE-38-22-E2-4B-03-F7
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Ethernet 5:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . : customer.com
   Description . . . . . . . . . . . : Check Point Virtual Network Adapter For Endpoint VPN Client
   Physical Address. . . . . . . . . : 54-DD-ED-17-0C-07
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Unknown adapter NETGEAR-VPN:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : TAP-Windows Adapter V9
   Physical Address. . . . . . . . . : 00-FF-9B-E5-70-08
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Wireless LAN adapter Local Area Connection* 2:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Microsoft Wi-Fi Direct Virtual Adapter #3
   Physical Address. . . . . . . . . : 94-E6-F7-53-59-77
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Wireless LAN adapter Local Area Connection* 3:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Microsoft Wi-Fi Direct Virtual Adapter #4
   Physical Address. . . . . . . . . : 96-E6-F7-53-59-76
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

Ethernet adapter Ethernet 2:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Fortinet Virtual Ethernet Adapter (NDIS 6.30)
   Physical Address. . . . . . . . . : 00-09-0F-FE-00-01
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Wireless LAN adapter Wi-Fi:

   Connection-specific DNS Suffix  . : local
   Description . . . . . . . . . . . : Intel(R) Wi-Fi 6 AX200 160MHz
   Physical Address. . . . . . . . . : 94-E6-F7-53-59-76
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   IPv4 Address. . . . . . . . . . . : 172.16.13.61(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.254.0
   Lease Obtained. . . . . . . . . . : Wednesday, March 30, 2022 9:09:47 AM
   Lease Expires . . . . . . . . . . : Saturday, April 2, 2022 9:09:56 AM
   Default Gateway . . . . . . . . . : 172.16.12.1
   DHCP Server . . . . . . . . . . . : 10.254.254.1
   DNS Servers . . . . . . . . . . . : 1.1.1.1
                                       1.1.1.1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Bluetooth Network Connection:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Bluetooth Device (Personal Area Network)
   Physical Address. . . . . . . . . : 94-E6-F7-53-59-7A
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
PS C:\WINDOWS\system32>`

Questions:
1) What are example commands to root cause the NIC binding from vNIC of WSL Ubuntu instance to physical NIC
2) Is there example power shell command where I can change binding from One Physical NIC to another (or back to original posting, bind it to "WSL" HyperV switch , or Windows Bridge.. and move that connection around to underlying "interface" that would allow the WSL instances to connect where I need.
3) Any insight if HyperV is required / used and hidden.   Vs Windows Bridge is the connector and packet path.

Thanks for responses

**Comment by penguinpages** [comment]
Update: 

After reboot , which I avoid like plague now due to the steps needed to get things back working with WSL, I have several weird things happen that may add needed color to above thread

1) The HyperV Virtual switch "WSL" linked to Gb NIC.. disapears
2) The Gb NIC has bridge checked as enabled.... but no icon showing "Network Bridge"
3) When I try to create HyperV switch called "WSL" external link to Gb NIC, it errors with code that it can't complete task that device with that name already exists.
4) When I try to launch WSL, it response:
~~~
PS C:\GitHub\cluster_devops\cluster_devops> wsl
The network was not found.
PS C:\GitHub\cluster_devops\cluster_devops>

What command can I run from powershell to query what Network is not found" that it is expecting?

**Comment by penguinpages** [comment]
So after windows update came out..  did my normal set aside block of 30 min post update to do the "get WSL networking to work again" dance.   As usual after reboot  the NIC devices "Bridge"(bridge device), and "WSL" (HyperV Virtual external switch with mgmt bound and set to bridge to Gb NIC) were missing.    Try routeine:

1) create new switch "WSL" in HyperV manager..  error that it "exists".    
2) Unbind HyperV and Bridge from Gb NIC (binding to ????) as neither exists.  THis gets basic OS packets working again.
3) Create "WLS" HyperV switch but only internal .. Success
4) Enable "WSL" swtich to attach to Gb NIC... errors...  HyperV and bridge show bindings again to NIC. remove again.
5) Attempt two: Enable "WSL" swtich to attach to Gb NIC... this time it takes.  OS now able to bridge mgmt over "WSL" HV switch
6) Bridge "WSL" and 'Ethernet" Gb NIC...  OS stops transmitting for a min but then picks back up.

But ... all above..  normal poking in dark to get things working.  But this time ... it is not working no matter which setups I try.   When start WSL services up  / restart along each step.. 
`PS C:\GitHub\penguinpages_cluster_devops\cluster_devops> ping 8.8.8.8 

Pinging 8.8.8.8 with 32 bytes of data:
Reply from 8.8.8.8: bytes=32 time=45ms TTL=113
Reply from 8.8.8.8: bytes=32 time=18ms TTL=113
Reply from 8.8.8.8: 
Ping statistics for 8.8.8.8:
    Packets: Sent = 3, Received = 2, Lost = 1 (33% loss),
bytes=32 Approximate round trip times in milli-seconds:
    Minimum = 18ms, Maximum = 45ms, Average = 21ms
Control-C
PS C:\GitHub\penguinpages_cluster_devops\cluster_devops> wsl
The network was not found.
PS C:\GitHub\penguinpages_cluster_devops\cluster_devops> `

Is there any registry key / ini file /  command to figure out what WSL thinks is binding adapter is using... so I can correct it?
This poking in the dark is not very helpful.   And I can't believe I am the only one here wondering how to root cause WSL -> Bridge? -> HyperV -> NIC   means to document and ensure things are configured correctly.

**Reply by elsaco** [reply]
@penguinpages why are you trying to manually manage WSL networking? Those are ephemeral ports. In fact only one `host network endpoint` is being created for WSL. If more than one WSL instance is running they all use the same IP, same route, etc. Even if you have a full Hyper-V installation WSL networking is managed by the host not end user.

**Comment by penguinpages** [comment]
I am not trying to manage them.  I am trying to get WSL to launch.   When WSL configures itself I think it creates a binding to the bridge.  But this bridge is not stable, goes away after reboot????  Not sure... but end result is  I can start WSL services, but cannot launch the WSL instance .. the only output is that the interface is missing.    What my post is asking is, what are the tools (registry keys, commands,  etc...) to map how the WSL VM instance and its NIC maps to the bridge which maps to the HV Vswitch which bridges to a physical interface.  In knowing that , I would hope to root cause why every reboot creates this discombobulated state.

**Reply by elsaco** [reply]
The WSL network interface is removed every time WSL shuts down, or the system reboots. You can watch the activity with `Get-WinEvent -ProviderName Microsoft-Windows-Hyper-V-VmSwitch` or `get-hnsendpoint`. I think the timeout is set to 8 sec. After that there's no output from `get-hnsendpoint` cmdlet. My only WSL peev is that the IP is different after each reset, and on Windows 10 I need to setup `xauth` and `$DISPLAY` every time the IP changes.

**Reply by penguinpages** [reply]
get-hnsendpoint     This is not on my Windows 11  system.  I am trying to poke around where that tool comes from as i ran across this article which has several very helpful pointers 

https://techcommunity.microsoft.com/t5/networking-blog/l2bridge-container-networking/ba-p/1180923

**Reply by penguinpages** [reply]
Just found where to download tool: https://www.powershellgallery.com/packages/HNS/0.2.4

'
Install-Module -Name HNS
'

**Comment by penguinpages** [comment]
Thanks... that was an important peice in that I now can see configuration events.  From the reboot this AM there is over 360... so I just attached a few to see if I can pick apart where WSL can no longer start, and how to remediate (as noted removing windows feature no longer is fixing it, nor is remove / adding "WSL" HV switch).


I am trying to find below litany of NIC and adapter IDs... within the registry or where I can see current vs what WSL references so maybe I can correct

`C:\WINDOWS\system32> Get-WinEvent -ProviderName Microsoft-Windows-Hyper-V-VmSwitch


   ProviderName: Microsoft-Windows-Hyper-V-VmSwitch

TimeCreated                      Id LevelDisplayName Message
-----------                      -- ---------------- -------
4/13/2022 1:09:13 PM            232 Information      NIC 9C49CC71-180E-480E-9434-B76BEA2ED675 (Friendly Name: Default Switch) successfully connected to port E7E8C081-7D10-4584-A6F4-751F36687B4F (Friendly Name: Container NIC 2ba129ba) on switch C08CB7B8-9B3C-408E-8E30...
4/13/2022 1:09:13 PM              7 Information      Miniport NIC 9C49CC71-180E-480E-9434-B76BEA2ED675 (Friendly Name: Default Switch) successfully initialized.
4/13/2022 1:09:13 PM            233 Information      The operation 'Create' succeeded on nic 9C49CC71-180E-480E-9434-B76BEA2ED675 (Friendly Name: Default Switch), Instance Id {00000000-0000-0000-0000-000000000000}.
4/13/2022 1:09:13 PM            264 Information      Port E7E8C081-7D10-4584-A6F4-751F36687B4F (Friendly Name: Container NIC 2ba129ba) successfully created on switch C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch).
4/13/2022 1:09:13 PM              9 Information      Switch C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch) successfully initialized.
4/13/2022 1:09:13 PM            232 Information      NIC C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch) successfully connected to port C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch) on switch C08CB7B8-9B3C-408E-8E30-5E16A3A...
4/13/2022 1:09:13 PM            233 Information      The operation 'Create' succeeded on nic C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch), Instance Id {00000000-0000-0000-0000-000000000000}.
4/13/2022 1:09:11 PM            233 Information      The operation 'Delete' succeeded on nic 4DF22C2F-53DB-4030-92EC-3A61DFDBF8B3 (Friendly Name: Default Switch), Instance Id {61b0aea8-f164-4d36-8956-f3a4d78a7fd9}.
4/13/2022 1:09:05 PM             62 Information      Switch C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch) successfully deleted.
4/13/2022 1:09:05 PM            233 Information      The operation 'Delete' succeeded on nic C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch), Instance Id {00000000-0000-0000-0000-000000000000}.
4/13/2022 1:09:05 PM            234 Information      NIC 4DF22C2F-53DB-4030-92EC-3A61DFDBF8B3 successfully disconnected from port 9C9325B1-E286-4707-9CAF-A8DDBE5A6DB6.
4/13/2022 1:09:05 PM            234 Information      NIC C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 successfully disconnected from port C08CB7B8-9B3C-408E-8E30-5E16A3AEB444.
4/13/2022 1:08:56 PM              7 Information      Miniport NIC 5405DB89-D138-4F70-BC17-2BF9AEA83F17 (Friendly Name: WSL) successfully initialized.
4/13/2022 1:08:53 PM              5 Information      Miniport NIC 5405DB89-D138-4F70-BC17-2BF9AEA83F17 (Friendly Name: WSL) successfully enabled
4/13/2022 1:08:53 PM            232 Information      NIC 5405DB89-D138-4F70-BC17-2BF9AEA83F17 (Friendly Name: WSL) successfully connected to port B95DAF2A-0108-4F08-B627-F733DCC79A51 (Friendly Name: WSL) on switch B54C7C62-1E03-4B5D-8626-5B7C90587913(Friendly Name: W...
4/13/2022 1:08:53 PM            233 Information      The operation 'Create' succeeded on nic 5405DB89-D138-4F70-BC17-2BF9AEA83F17 (Friendly Name: WSL), Instance Id {00000000-0000-0000-0000-000000000000}.
4/13/2022 1:08:53 PM              5 Information      Miniport NIC 4DF22C2F-53DB-4030-92EC-3A61DFDBF8B3 (Friendly Name: Default Switch) successfully enabled
4/13/2022 1:08:53 PM            232 Information      NIC 4DF22C2F-53DB-4030-92EC-3A61DFDBF8B3 (Friendly Name: Default Switch) successfully connected to port 9C9325B1-E286-4707-9CAF-A8DDBE5A6DB6 (Friendly Name: Container NIC 7922a227) on switch C08CB7B8-9B3C-408E-8E30...
4/13/2022 1:08:53 PM              7 Information      Miniport NIC 4DF22C2F-53DB-4030-92EC-3A61DFDBF8B3 (Friendly Name: Default Switch) successfully initialized.
4/13/2022 1:08:53 PM            233 Information      The operation 'Create' succeeded on nic 4DF22C2F-53DB-4030-92EC-3A61DFDBF8B3 (Friendly Name: Default Switch), Instance Id {00000000-0000-0000-0000-000000000000}.
4/13/2022 1:08:53 PM            264 Information      Port 9C9325B1-E286-4707-9CAF-A8DDBE5A6DB6 (Friendly Name: Container NIC 7922a227) successfully created on switch C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch).
4/13/2022 1:08:53 PM              9 Information      Switch C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch) successfully initialized.
4/13/2022 1:08:53 PM            232 Information      NIC C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch) successfully connected to port C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch) on switch C08CB7B8-9B3C-408E-8E30-5E16A3A...
4/13/2022 1:08:53 PM            233 Information      The operation 'Create' succeeded on nic C08CB7B8-9B3C-408E-8E30-5E16A3AEB444 (Friendly Name: Default Switch), Instance Id {00000000-0000-0000-0000-000000000000}.
4/13/2022 1:08:53 PM            264 Information      Port B95DAF2A-0108-4F08-B627-F733DCC79A51 (Friendly Name: WSL) successfully created on switch B54C7C62-1E03-4B5D-8626-5B7C90587913 (Friendly Name: WSL).
4/13/2022 1:08:53 PM            264 Information      Port 284AF3FD-0785-473D-8CE2-22C65343CAF9 (Friendly Name: WSL_External) successfully created on switch B54C7C62-1E03-4B5D-8626-5B7C90587913 (Friendly Name: WSL).
4/13/2022 1:08:53 PM              9 Information      Switch B54C7C62-1E03-4B5D-8626-5B7C90587913 (Friendly Name: WSL) successfully initialized.
4/13/2022 1:08:53 PM            232 Information      NIC B54C7C62-1E03-4B5D-8626-5B7C90587913 (Friendly Name: WSL) successfully connected to port B54C7C62-1E03-4B5D-8626-5B7C90587913 (Friendly Name: WSL) on switch B54C7C62-1E03-4B5D-8626-5B7C90587913(Friendly Name: W...
4/13/2022 1:08:53 PM            233 Information      The operation 'Create' succeeded on nic B54C7C62-1E03-4B5D-8626-5B7C90587913 (Friendly Name: WSL), Instance Id {00000000-0000-0000-0000-000000000000}.
4/13/2022 1:08:53 PM            277 Information      vmswitch.sys build 22000.amd64.co_release_svc_prod1, debug false, official true, 1 1`

**Reply by elsaco** [reply]
There's also `hnsdiag.exe` what will show WSL networking information. Type `hnsdiag /?` to see usage.

**Comment by penguinpages** [comment]
I tried to make heads or tails of that tool output. 

I did a full removal of "ubuntu" and WSL (windows feature).  Did a reinstall ..  and ..  I still have consistent error:

'
Installing, this may take a few minutes...
WslRegisterDistribution failed with error: 0x803b0001
Error: 0x803b0001 The network was not found.

Press any key to continue...


'

Attached is output from file.  
[wsl_error_hnsdiag.txt](https://github.com/microsoft/WSL/files/8508054/wsl_error_hnsdiag.txt)

The tool lacks any help and I googled around and did not find anything to help.  All the devices it lists...  are these the current set of NICs... (VPN , physical etc..) or is this just a bunch of orphaned WSL ones never cleaned up.

Documentation on tool :  https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/common-problems

My notes:
HyperV Switch Name: "WSL"
HyperV Switch Named "WSL" ID: 2A059187-2410-4316-95F9-7B2E07E38D17

The rest of the file is a bunch of  Partition ID with associated what seems to be almost DMA / NIC ID to bridge

Ex:'
OwningPartitionId : e0e16197-dd56-4a10-9195-5ee7a155a838 (Host Partition)  ---> My guess is this is the "Ubuntu" ID
TargetPartitionId : e0e16197-dd56-4a10-9195-5ee7a155a838 (Host Partition)  ---> ID same as owning ID, maybe  WSL uses this for VM.
    Protocol         : UDP   ----> Packet type for NAT????
    DeliveryMode     : DatapathDeliveryIgnore  ----> No idea
    ReservationType  : Exclusion  ----> No idea
    SharingFlags     : Exclusion UnicastSpecificBind NonunicastSpecificBind    ----> Maybe for NAT tuning
    Priority         : 0x1ffffffffffff84    ----> boot or NIC slot notation
    StartingPort     : 64821    ----> Maybe this is like a NIC/PCI DMA range set .. or NIC port in virtual switch / virtual bridge port number?
    EndingPort       : 64920


My guess: I need to figure out.. via some other means.. when WSL loads up.. what "NIC" it is trying to bind to. And then find that key in the registry and change it for current  bridge that binds to HyperV switch "WSL"

**Comment by penguinpages** [comment]
Came back to poke at this topic.    I think the concept of the bridge overlay on HV Switch, is kind of fubar.

I think the VFP mode on NIC is what WSL and HyperV need for bridging packets.   I need to of course survive a few reboots and maybe even an update from Microsoft :)  before I can set this as "root cause"
![WSL_VFP](https://user-images.githubusercontent.com/17153876/165204423-a7286e64-470b-4b1b-a784-bd9a43b93aff.png)

**Reply by penguinpages** [reply]
<Yet another update>

Nope... this does not fix the mess that is Windows , HyperV and WSL.   Update last night... apply.. network died.. reboot.   "WSL" external hyperV switch missing.  Binding to Gb NIC still was left over.   Could not create new switch named "WSL" to wifi NIC (baseline of can it create with same name and bind to another NIC).  Had to remove "bridge" and "hyperv" from Gb nic to create a new one.  IP then worked (works) fine for Windows OS and HyperV VM, but WSL back to the same FUBAR.

'
PS C:\GitHub\\cluster_devops> wsl
The network was not found.
'

I will poke some more to try to create repair baseline.  it would be very helpful though if instead of random "add / delete"  "stand on one foot... turn twice...  sprinkle holy coffee on computer,... recreate bridge... WOW...its working again.. DON"T TOUCH" ,  I could just use my system and trust that a reboot/patch, will not cost me 30min.



![WSL_HV_Disapears](https://user-images.githubusercontent.com/17153876/165745371-d110c7a6-051f-4770-87f6-af0667c605bd.png)
